### PR TITLE
Turn serializer enums into enum classes.

### PR DESCRIPTION
**Pull request by AlCutter**
Change the serializer enums into enum classes (thanks to @pphaneuf  for slicing this out of my monster serialization PR!)

### CODE REVIEW

File: cpp/log/log_signer.cc [hunk 1]
```diff
@@ -136,26 +133,25 @@ LogSigner::SignResult LogSigner::SignTreeHead(SignedTreeHead* sth) const {
 }
 
 // static
-LogSigner::SignResult LogSigner::GetSerializeError(
-    Serializer::SerializeResult result) {
+LogSigner::SignResult LogSigner::GetSerializeError(SerializeResult result) {
   SignResult sign_result;
   switch (result) {
-    case Serializer::INVALID_ENTRY_TYPE:
+    case SerializeResult::INVALID_ENTRY_TYPE:
```
Comment (Martin2112) [inline cpp/log/log_signer.cc]: Manually mapping enum values like this seems a bit messy. Can defer fixing it to a later PR.

File: cpp/log/log_signer.cc [hunk 2]
```diff
@@ -252,34 +248,34 @@ LogSigVerifier::VerifyResult LogSigVerifier::VerifySTHSignature(
...
 }
 
 // static
 LogSigVerifier::VerifyResult LogSigVerifier::GetSerializeError(
-    Serializer::SerializeResult result) {
+    SerializeResult result) {
   VerifyResult verify_result;
   switch (result) {
-    case Serializer::INVALID_ENTRY_TYPE:
+    case SerializeResult::INVALID_ENTRY_TYPE:
```
Comment (Martin2112) [inline cpp/log/log_signer.cc]: Again I think we have a proliferation of enums that represent very similar things. It's OK for now.

File: cpp/proto/serializer.h [hunk 3]
```diff
@@ -10,34 +10,57 @@
...
+  EXTENSIONS_TOO_LONG,
+  INVALID_KEYID_LENGTH,
+  EMPTY_LIST,
+  EMPTY_ELEM_IN_LIST,
+  LIST_ELEM_TOO_LONG,
+  LIST_TOO_LONG,
+  EXTENSIONS_NOT_ORDERED,
+};
+
+std::ostream& operator<<(std::ostream& stream, const SerializeResult& r);
```
Comment (Martin2112) [inline cpp/proto/serializer.h]: Did you pull these operators into the PR?

Comment (AlCutter) [inline cpp/proto/serializer.h]: They're present, but github seems not to want to show the diffs in `serializer.cc`, it says:
"341 additions, 280 deletions not shown" for that file, not entirely sure why...

**Comment by Martin2112** [comment]
You're right, it's not showing.

**Comment by AlCutter** [comment]
Pushed it to Reitveld to see if that'll do a better job: http://codereview.appspot.com/271700043

**Comment by Martin2112** [comment]
Reviewed on Rietveld. Looks good.

**Pull request merged**